clang/tools/driver/cc1depscanProtocol.cpp

```diff
@@ -169,11 +170,28 @@ Expected<ScanDaemon> ScanDaemon::launchDaemon(StringRef BasePath,
   // Only do it the first time.
   LaunchTestDaemon = false;
 
+  // Spawn attributes
+  posix_spawnattr_t attrs;
+  if (int EC = posix_spawnattr_init(&attrs))
+    return llvm::errorCodeToError(std::error_code(EC, std::generic_category()));
+  auto attrs_cleanup =
+      llvm::make_scope_exit([&] { posix_spawnattr_destroy(&attrs); });
+
+#ifdef POSIX_SPAWN_CLOEXEC_DEFAULT
+  // In the spawned process, close all file descriptors that are not explicitly
+  // described by the file actions object. This is particularly relevant for
+  // llbuild which waits on a "control file handle" and, if inherited, it would
+  // cause llbuild to wait for the spawned process to exit.
+  // FIXME: This is Darwin-specific extension, perform the same function on
+  // non-darwin platforms.
+  if (int EC = posix_spawnattr_setflags(&attrs, POSIX_SPAWN_CLOEXEC_DEFAULT))
+    return llvm::errorCodeToError(std::error_code(EC, std::generic_category()));
+#endif
+
   ::pid_t Pid;
-  int EC =
-      ::posix_spawn(&Pid, Args[0], /*file_actions=*/nullptr,
-                    /*attrp=*/nullptr, const_cast<char **>(LaunchArgs.data()),
-                    /*envp=*/nullptr);
+  int EC = ::posix_spawn(&Pid, Args[0], /*file_actions=*/nullptr, &attrs,
```

Comment:
I wonder if calling with `attrs == nullptr` is the same as an empty attrs. From documentation, `nullptr` should inherit everything from the parent(?) while the empty set might reset the attribute? Not sure but maybe someone would know the detail.

Reply:
From the `posix_spawnattr_init` manpage, it seems like the default / empty set is likely to match the parent:
```
     The posix_spawnattr_init() function initializes the object referenced by
     attr, to an empty set of spawn attributes for subsequent use in a call to
     posix_spawn(2) or posix_spawnp(2).

     A posix_spawnattr_t object is an opaque type that is used to collect set of
     spawn attributes specified by calls to posix_spawnattr_setflags(3),
     posix_spawnattr_setpgroup(3), posix_spawnattr_setsigdefault(3),
     posix_spawnattr_setsigmask(3), posix_spawnattr_setarchpref_np(3),
     posix_spawnattr_setbinpref_np(3), posix_spawnattr_setspecialport_np(3), or
     posix_spawnattr_setexceptionports_np(3) such that once the spawn operation
     is complete, the child process will have a different set of attributes than
     the parent.  The order of these operations is unspecified and may change;
     programs should not depend on their ordering.
```
"[...] such that once the spawn operation is complete, the child process will have a different set of attributes than the parent"

My reading is that this is fine as-is.

Reply:
My reading is same as @dexonsmith; from what I understand `posix_spawn` has default behavior which can be overridden using user specified spawn attributes, and whether you pass `NULL` or an empty set it has the same effect.